File: test/dummy/app/models/article.rb
# frozen_string_literal: true

class Article < ApplicationRecord
  CATEGORIES = ['Arts and Entertainment', 'Cars & Other Vehicles', 'Computers and Electronic',
                'Education and Communications', 'Family Life', 'Finance and Business',
                'Food and Entertaining', 'Health', 'Hobbies and Crafts',
                'Holidays and Traditions', 'Home and Garden', 'Pets and Animals',
                'Travel'].freeze

  has_many :comments, dependent: :destroy
  belongs_to :user

  attr_reader :published
  attr_accessor :admin_review_started_at

  validates :title, presence: true, length: { minimum: 3, maximum: 255 }
  validates :text, presence: true, length: { minimum: 100 }

  validate :vulgarity_level
  before_update :calculate_admin_review_time

  class << self
    def published(state = true)
      if state
        where.not published_at: nil
      else
        where published_at: nil
      end
    end
  end

  def published=(val)
    @published = case val.class.name
                 when 'String' then val == '1'
                 when 'NilClass' then nil
                 else val ? true : false
                 end
    if published
      set_published_at
    else
      self.published_at = nil
    end
  end

  def published?
    published_at.present?
  end

  def publish
    set_published_at
    changed? ? save : nil
  end

  def destroy(force = false)
    return false if published? && !force

    super()
  end

  private

  def vulgarity_level
    return unless contains_vulgarity?(title) || contains_vulgarity?(text)

    errors.add :base, 'Article contains strong language.'
  end

  def contains_vulgarity?(attrib)
    vulgar_word = 'fuck'
    attrib.present? && attrib =~ /#{vulgar_word}/i
  end

  def set_published_at
    return nil if published_at.present?

    self.published_at = Time.current
  end

  def calculate_admin_review_time
    return if admin_review_started_at.nil?

    val = (Time.now.to_f * 1000 - admin_review_started_at.to_f) / 1000
    self.admin_review_time = val.round 2
  end
end

File: test/dummy/frontend/js/models/Article.js
import { Models } from "loco-js";

class Article extends Models.Base {
  static identity = "Article";

  static resources = {
    url: "/user/articles",
    paginate: { per: 5 },
    main: {
      url: "/articles",
      paginate: { per: 3 }
    },
    admin: {
      url: "/admin/articles",
      paginate: { per: 4 }
    }
  };

  static attributes = {
    title: {
      validations: {
        presence: true,
        length: { within: [3, 255] }
      }
    },
    content: {
      validations: {
        presence: true,
        length: { minimum: 100 }
      },
      remoteName: "text"
    },
    createdAt: {
      type: "Date",
      remoteName: "created_at"
    },
    updatedAt: {
      type: "Date",
      remoteName: "updated_at"
    },
    commentsCount: {
      type: "Int",
      remoteName: "comments_count"
    },
    publishedAt: {
      type: "Date",
      remoteName: "published_at"
    },
    published: {},
    adminReview: {
      remoteName: "admin_review"
    },
    adminRate: {
      type: "Int",
      remoteName: "admin_rate"
    },
    categoryId: {
      type: "Int",
      remoteName: "category_id"
    },
    adminReviewStartedAt: {
      remoteName: "admin_review_started_at"
    }
  };

  static validate = ["vulgarityLevel"];

  constructor(data) {
    super(data);
    this.published = this.publishedAt == null ? false : true;
  }

  vulgarityLevel() {
    if (
      (this.title != null && /fuck/i.exec(this.title)) ||
      (this.content != null && /fuck/i.exec(this.content))
    )
      this.addErrorMessage("Article contains strong language.", {
        for: "base"
      });
  }

  setDefaultValuesForAdminReview() {
    this.adminRate = this.adminRate == null ? 3 : this.adminRate;
    this.categoryId = this.categoryId == null ? 6 : this.categoryId;
    this.adminReviewStartedAt = Date.now();
  }
}

export default Article;
